File: src/leetcode/LongestCommonSubsequence.java
package leetcode;

import java.util.Arrays;

/**
 * https://github.com/mission-peace/interview/blob/master/src/com/interview/dynamic/LongestCommonSubsequence.java
 */
public class LongestCommonSubsequence {

    static int longestCommonSubsequence(String text1, String text2) {

        int max = 0;
        int[][] dp = new int[text1.length()][text2.length()];
        for (int[] ints : dp) {
            Arrays.fill(ints, 0);
        }

        for (int i = 0; i < text1.toCharArray().length; i++) {
            for (int j = 0; j < text2.toCharArray().length; j++) {
                if (text1.charAt(i) == text2.charAt(j)) {
                    if (i - 1 >= 0 && j - 1 >= 0) {
                        dp[i][j] = dp[i - 1][j - 1] + 1;
                    } else {
                        dp[i][j] = 1;
                    }
                } else {
                    if (i - 1 >= 0 && j - 1 >= 0) {
                        dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                    } else if (i - 1 >= 0) {
                        dp[i][j] = dp[i - 1][j];
                    } else if (j - 1 >= 0) {
                        dp[i][j] = dp[i][j - 1];
                    } else {
                        // We can also skip this since the initialization of dp already initializes all cells to 0
                        dp[i][j] = 0;
                    }
                }
                max = Math.max(dp[i][j], max);
            }
        }

        return max;
    }

    public static void main(String[] args) {
        System.out.println(longestCommonSubsequence("abcde","ace"));
    }
}

File: python-code/longest_common_subsequence.py
class Solution:
    def __init__(self):
        self.longest_length: int = 0

    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        if text1 == text2:
            return len(text1)

        self.traverse(text1, text2, 0)

        return self.longest_length

    def traverse(self, a: str, b: str, local_longest: int):
        """
        abc
        bc
        t(abc,c)
            t(abc,None)
                return with no side effects
            t(bc,c)
                t(c,c) ll=1
                    t(None,None)
                t(bc,None)
                    return with no side effects
        t(bc,bc)
            t(c,c)
                t(None,None)
        """

        if not a and not b:
            self.longest_length = max(self.longest_length, local_longest)

        if not a or not b:
            return

        if a[0] == b[0]:
            self.traverse(a[1:], b[1:], local_longest + 1)
            return

        self.traverse(a, b[1:], local_longest)
        self.traverse(a[1:], b, local_longest)


if __name__ == "__main__":
    print(Solution().longestCommonSubsequence("abcde", "ace"))
    print(Solution().longestCommonSubsequence("abc", "abc"))
    print(Solution().longestCommonSubsequence("abc", "def"))
